include/range/v3/view/drop_last.hpp

```diff
@@ -0,0 +1,159 @@
+/// \file
+// Range v3 library
+//
+//  Copyright Andrey Diduh 2019
+//
+//  Use, modification and distribution is subject to the
+//  Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at
+//  http://www.boost.org/LICENSE_1_0.txt)
+//
+// Project home: https://github.com/ericniebler/range-v3
+//
+
+#ifndef RANGES_V3_VIEW_DROP_LAST_HPP
+#define RANGES_V3_VIEW_DROP_LAST_HPP
+
+#include <type_traits>
+#include <meta/meta.hpp>
+#include <range/v3/range_concepts.hpp>
+#include <range/v3/view/all.hpp>
+#include <range/v3/view/view.hpp>
+#include <range/v3/view_interface.hpp>
+#include <range/v3/view_adaptor.hpp>
+#include <range/v3/utility/functional.hpp>
+
+namespace ranges
+{
+    inline namespace v3
+    {
+        template<typename Rng>
+        struct drop_last_bidirectional_view
+          : view_interface<drop_last_bidirectional_view<Rng>, range_cardinality<Rng>::value>
+        {
+            CONCEPT_ASSERT(ranges::BidirectionalRange<Rng>());
+        private:
+            friend range_access;
+            Rng rng;
+            std::size_t n;
+            detail::non_propagating_cache<sentinel_t<Rng>> sentinel;
+            void update_sentinel()
+            {
+                if (!sentinel)
+                {
+                    sentinel = ranges::prev(ranges::end(rng), n, ranges::begin(rng));
+                }
+            }
+        public:
+            drop_last_bidirectional_view() = default;
+            drop_last_bidirectional_view(Rng&& rng, std::size_t n)
+              : rng(std::move(rng)), n(n)
+            {}
+
+            iterator_t<Rng> begin()
+            {
+                return ranges::begin(rng);
+            }
+            sentinel_t<Rng> end()
+            {
+                update_sentinel();
+                return *sentinel;
+            }
```

Comment:
You'll need `const`-qualified `begin`/`end` also.

Reply:
I can't - because of cached sentinel. Same as with `view::filter` .

Reply:
From the other side - I'll just add on the fly computation for RandomAccessRange const end .